### DISCUSSION TITLE
Run a queue every 3 months in a specified timezone?

### DISCUSSION THREAD

**Discussion by deadcoder0904** [quirrel-dev/quirrel#855] [Q&A] [answered]
First of all, thanks for Quirrel. It is really easy to use.

I want to run a queue every 3 months in a specified timezone. I checked the API & `CronJob` has a timezone.

I also checked [Queue](https://docs.quirrel.dev/api/queue#enqueue) & saw it has a `cron` property. But I'm confused. Do I need to use `every` or `times` when I use `cron` with `schedule` & `timezone`?

Basically, I want to send a queue every 3 months including right now in a specific timezone like `Asia/Kolkata`. How do I do it?

Currently, my code looks like:

```ts
await hello.enqueue(name, {
    id: name,
    delay: '30sec',
    repeat: {
      every: '10sec',
      times: 2, // 2 * 10sec = 20sec
    },
  })
```

I have made a sample demo repo → https://github.com/deadcoder0904/quirrel-log

Right now, with my code, it gave the scheduled date many hours later in the quirrel ui but executed it properly. I guess that's because of the timezone issue.

Also, is there something like `Infinity` where cron runs for eternity, or do we have to specify a big number like 1000 or so?

**Comment by Skn0tt** [comment, marked as answer]
Hi Akshay! `CronJob` is meant for "static" cronjobs that are known at development-time. Things like "purge stale data from the database" or "send out some recurring email to customers".

If you want to dynamically schedule a cron job, you can do so using `Queue`'s `repeat.cron` property.
Imagine you want to send out an end-of-week-summary to every user on your platform, but every user has their own time. Here's a snippet for that:

```ts
const user = {
  id: "abc",
  summarySchedule: "15_14_*_*_5", // every friday at 14:15, see: https://crontab.guru/#15_14_*_*_5
  timezone: "Asia/Kolkata",
  // ... potentially other user properties ...
}

await summaryQueue.enqueue(user.id, {
    id: user.id,
    repeat: {
      cron: [ user.summarySchedule, timezone ],
    }
})
```

Because this user is not known at development time, we can't define a `CronJob` for them. We also wouldn't wanna do this for every single user 😅. That's why we use a regular `Queue`. We specify the `id`, because this allows us to clean up the job when it's not needed anymore:

```ts
async function unsubscribeUserFromSummary(userId: string) {
  await summaryQueue.delete(userId)
}
```

The `times` field can be used to limit the job from repeating a maximum number of `times`. So `repeat: { every: "5sec", times: 4 }` means 4 executions, 5 seconds apart. If you leave out the `times`, the job will repeat indefinitely, or until you delete it. You can combine `times` with `cron` in the same way.
